api/src/main/java/org/openmrs/module/registrationcore/api/impl/PatientCreatedListener.java

```diff
@@ -0,0 +1,108 @@
+    private MpiProperties mpiProperties;
+
+    private IdentifierBuilder identifierBuilder;
+
+    public void setMpiProperties(MpiProperties mpiProperties) {
+        this.mpiProperties = mpiProperties;
+    }
+
+    public void setIdentifierBuilder(IdentifierBuilder identifierBuilder) {
+        this.identifierBuilder = identifierBuilder;
+    }
+
+    /**
+     * @return a list of Actions this listener can deal with
+     */
+    public List<String> subscribeToActions(){
+        List actions = new ArrayList<String>();
+        actions.add(Event.Action.CREATED.name());
+        return actions;
+    }
+
+    /**
+     * Export patient to MPI server & update local
+     * patient with new one patient identifier.
+     *
+     * @param message message with properties.
+     */
+    @Override
+    public void performMpiAction(Message message) {
+        Patient patient = extractPatient(message);
+        if (patient != null){
+            PatientIdentifierType mpiPatientIdType = patientService.getPatientIdentifierTypeByUuid(mpiProperties.getMpiPersonIdentifierTypeUuid());
+            PatientIdentifier mpiPatientId = patient.getPatientIdentifier(mpiPatientIdType);
+            if (mpiPatientId != null && !mpiPatientId.getIdentifier().isEmpty()){
+                // Patient originated in MPI and thus should not be pushed to the MPI
+                return;
+            }else{
+                String mpiPatientIdString = null;
+                try {
+                    mpiPatientIdString = pushPatientToMpiServer(patient);
+                } catch (Exception e) {
+                    ErrorHandlingService errorHandler = coreProperties.getPixErrorHandlingService();
+                    if (errorHandler == null) {
+                        throw new MpiException("PIX patient push exception occurred "
+                                + "with not configured PIX error handler", e);
+                    } else {
+                        errorHandler.handle(prepareParameters(patient),
+                                PixErrorHandlingService.SENDING_PATIENT_AFTER_PATIENT_CREATION_DESTINATION,
+                                true,
+                                ExceptionUtils.getFullStackTrace(e));
+                        throw new MpiException("PIX patient push exception occurred", e);
+                    }
+                }
+                if (mpiPatientIdString != null) {
+                    updatePatient(patient, mpiPatientIdString);
+                }
+            }
+        }else{
+            throw new MpiException("PeformMpiAction error: extractPatient returned null patient");
+        }
+    }
+
+    private String pushPatientToMpiServer(Patient patient) {
+        return coreProperties.getMpiProvider().exportPatient(patient);
+    }
+
+    private void updatePatient(final Patient patient, final String mpiPatientId) {
+        User creator = patient.getCreator();
+        patient.addIdentifier(createPatientIdentifier(creator, mpiPatientId));
```

Comment:
Do you really need to add the mpi id to the patient in the local instance? If you know the patient id in the local instance, I'd expect that you push it to the MPI , if you need to push updates to the MPI, you should identify the, by the local id

Reply:
Was there any design discussions around this functionality?

Reply:
Hmmm we did not have explicit discussions about this, although this question was brought up by @mogoodrich at one point. We can have the discussion now I guess... There might not be an explicit function for this, like what you said the system should be able to work without it, but might it add robustness/ improve the integrity of the records? @approce, might you be able to comment on why you designed it in this way? @pgesek, any thoughts?

Reply:
Do you wanna discuss this from Talk?

Reply:
Ok moved to Talk, [over here](https://talk.openmrs.org/t/local-storage-of-the-master-patient-index-mpi-identifier/19216)